### PR TITLE
cast: eip1967 commands

### PR DESCRIPTION
**Pull request by tynes**
## Motivation

I want an easy to to tell if a contract is a proxy and to be able to know the proxy admin and proxy impl using `cast`.

## Solution

Implements two new commands for `cast` that fetch common EIP-1967 storage slots and print the values as address strings.

```
$ cast admin
$ cast implementation
```

This makes it very easy to determine if a contract is proxied.

Closes https://github.com/foundry-rs/foundry/issues/4168

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
great!

lgtm

File: cast/src/lib.rs [hunk 1]
```diff
@@ -491,6 +491,64 @@ where
...
+    /// # }
+    /// ```
+    pub async fn implementation<T: Into<NameOrAddress> + Send + Sync>(
+        &self,
+        who: T,
+        block: Option<BlockId>,
+    ) -> Result<String> {
+        let slot =
+            H256::from_str("0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc")?;
+        let value = self.provider.get_storage_at(who, slot, block).await?;
```
Comment (mds1) [inline cast/src/lib.rs:518]: Should this call a contract's `function implementation() returns (address)` method first, then query the EIP-1967 slot if that call fails? That way it supports non-standard proxies too that don't follow EIP-1967, with compound and compound governors being the common ones

Comment (mds1) [reply cast/src/lib.rs:518]: Calling `implementation()` second, if the 1967 slot is empty, might be better actually

File: cast/src/lib.rs [hunk 2]
```diff
@@ -491,6 +491,64 @@ where
...
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn admin<T: Into<NameOrAddress> + Send + Sync>(
+        &self,
+        who: T,
+        block: Option<BlockId>,
+    ) -> Result<String> {
+        let slot =
+            H256::from_str("0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103")?;
```
Comment (mds1) [inline cast/src/lib.rs:546]: Similarly, should this fall back to trying `admin()` if nothing is found?

File: cast/src/lib.rs (hunk 1 above)
Comment (tynes) [reply cast/src/lib.rs:518]: Definitely in favor if this in a follow up PR

Comment (DaniPopes) [reply cast/src/lib.rs:518]: There's also an `implementation` field in the etherscan `contracts` route: https://docs.rs/ethers/latest/ethers/etherscan/contract/struct.Metadata.html

**Pull request merged**